### ISSUE TITLE
Only load if package is available

### ISSUE THREAD

**Issue by Vurp** [jwiegley/use-package#739]
It seems that packages will try to load even if it doesn't exist. This occurs even when `use-package-check-before-init` is set to `t`. For example the following will crash on startup if lispyville is not installed:

```
(use-package lispyville
  :diminish (lispyville-mode-line-string " 🍰")
  :hook ((clojure-mode emacs-lisp-mode lisp-mode) . lispyville-mode))
```

It could be worked around by wrapping everything inside a `(with-eval-after-load ...)` form, but I prefer to have a way of doing it universally in all use-package definitions to prevent copy-pasting.

**Comment by noctuid** [comment]
`use-package-check-before-init` only works for `:init`. It would be nice to be able to control which keywords can run if the package is not available.

**Comment by noctuid** [comment]
Also, this is basically the same as #693 (except that issue may have been specifically for `:config`; either way, I think only one open issue is needed). What you can do for now is use `:if`, but this requires using a wrapper around `use-package` if you want to do it automatically.

**Comment by Vurp** [comment]
How could I define such a wrapper?

**Comment by noctuid** [comment]
Something like this:

```elisp
(defmacro my-use-package (name &rest args)
  (declare (indent 1))
  `(use-package ,name
     :when (package-installed-p ',name)
     ,@args))
```

**Comment by waymondo** [comment]
Yeah that's a good solution. You could also register that default `:when` handler to `use-package-defaults` as well.

**Comment by noctuid** [comment]
That's a much better idea. I didn't realize you can have use-package still keep a default value for a keyword even when the keyword is manually specified with `use-package-defaults`. I personally use `locate-library` instead since I'm not using package.el. This seems to work well:

```elisp
(cl-pushnew '(:when
              (lambda (package &rest _)
                `(locate-library (symbol-name ',package)))
              t)
            use-package-defaults
            :test #'equal)
```

**Comment by skangas** [comment]
It seems like this is resolved by using `:if (locate-library "foo.el")`. I'm therefore closing this issue.

**Comment by xeruf** [comment]
Would be good to add it to the docs though, as this was not easy to find for me.

**Comment by skangas** [comment]
> Would be good to add it to the docs though, as this was not easy to find for me.

Did you install this package using `package.el`, or did you use some other method? If you used some other method, could you explain what you did?

I'm trying to understand the use-case here, and if `locate-library` could perhaps be avoided somehow (because it is slow).

**Comment by xeruf** [comment]
With rdictcc as well as lilypond, the el-package is delivered with the system package, so I want to check for that to be available before applying the package configuration.

**Comment by skangas** [comment]
> With rdictcc as well as lilypond, the el-package is delivered with the system package, so I want to check for that to be available before applying the package configuration.

Thanks, I'm reopening the bug.
